Feature/payment

sangkim98 opened the pull request:
payment and payment manager are nearly complete.

sangkim98 committed:
quick implementation

sangkim98 committed:
quick implementation

sangkim98 committed:
throwing of exceptions cleared and work in progress

yeohongwei committed:
Initial draft for RoomServiceOrder and MenuItem not for merging but for discussion

sangkim98 committed:
added javadoc

sangkim98 committed:
merged with main

sangkim98 committed:
generate and print bill partially done

sangkim98 committed:
added javadoc and AddNewObject implemented

sangkim98 committed:
completed payment and manager class and Supermanager edit

sangkim98 committed:
reverted changes in SearchList

sangkim98 committed:
search list by room number

sangkim98 committed:
re-added guest ID as payment attribute

sangkim98 committed:
revert changes to menu items and room service

sangkim98 committed:
payment complete

stephwongwt commented on src/main/payment/main/G1HRPS/Payment.java:
```diff
@@ -1,164 +1,277 @@
-	public void SetDiscount(float discount) {
-		// TODO - implement Payment.SetDiscount
+	/**
+	 * 
```
Add a brief here and for all other getters. e.g. Retrieves guest id of this guest. And these should end with a period.

stephwongwt commented on src/main/payment/main/G1HRPS/Payment.java:
```diff
@@ -1,164 +1,277 @@
 	}
 
 	/**
+	 * Assigns new discount rate to this payment
```
Are we using discount rate (i.e. 10%) or using discount value(i.e. $5)? Also how should this value be used? To be automatically deducted from bill total? Best to specify.

sangkim98 replied:
I'm assuming we are going to enter discount rate before the bill is to be generated and later applied in the calculation of `GenerateAndPrintBill()`. Haven't specified the format, so I'll clarify that!

stephwongwt commented on src/main/payment/main/G1HRPS/Payment.java:
```diff
@@ -1,164 +1,277 @@
+	 */
 
-		return guest_id_;
+	public void SetRsCharges(float rs_charges) {
```
Please rename the method such that it is clear what you're setting. i.e. SetRoomServiceCharges

stephwongwt commented on src/main/payment/main/G1HRPS/Payment.java:
```diff
@@ -1,164 +1,277 @@
 	private float discounts_;
 	private float tax_;
+	private float room_charges_;
+	private float rs_charges_;
```
Please use clearly understandable naming for variables, i.e. room_service_charges_

sangkim98 replied:
I'll resolve the getters description and unclear var/method names

stephwongwt commented on src/main/payment/main/G1HRPS/Payment.java:
```diff
@@ -1,164 +1,277 @@
+	 */
 
-		return this.discounts_;
+	public void SetRsCharges(List<RoomServiceOrder> room_service_order_list) {
```
I don't think this is very efficient in the sense that if there are a lot of room services from many different guests and rooms, each time you want to calculate this, you have to go through the whole list. The provided list should already be filtered to relate to only this guest, looking at our miro board, we search for a guest to check out, then print bill info for said guest.

This can be just a float value. RoomServiceManager should have a getlist function that provides a list ofRoom Service Orders by guest.

stephwongwt commented on src/main/payment/main/G1HRPS/Payment.java:
```diff
@@ -1,164 +1,277 @@
+	 * @return
+	 */
+
 	public float SetBillTotal() {
```
This should be called CalculateBillTotal

stephwongwt commented on src/main/payment/main/G1HRPS/PaymentManager.java:
```diff
@@ -15,32 +15,194 @@
+	 */
+
+	@Override
+	public Payment SearchList(String key) {
```
parameter name should be payment_id as listed in the @param documentation

stephwongwt commented on src/main/payment/main/G1HRPS/PaymentManager.java:
```diff
@@ -15,32 +15,194 @@
+		System.out.println("-- Cost of Stay : " + payment.GetRoomCharges());
+		System.out.println("- Room services ordered");
+		for(RoomServiceOrder rso : rs_order_list) {
+			if(rso.GetRoomNum() == room_num) {
```
This section here looks like it's duplicated at your GetRsCharges method.

stephwongwt commented on src/main/payment/main/G1HRPS/PaymentManager.java:
```diff
@@ -15,32 +15,194 @@
+	 * @param days_of_stay
+	 * @param rs_order_list
+	 */
+	public void GenerateAndPrintBill(Payment payment, int days_of_stay, List<RoomServiceOrder> rs_order_list) {
```
GenerateAndPrintBill should be the one that creates the Payment object.

notingcode replied:
I'll just rename the method to `GenerateBill()` and override `toString()` for Payment class.

stephwongwt commented on src/main/payment/main/G1HRPS/PaymentManager.java:
```diff
@@ -15,32 +15,194 @@
+		System.out.println("-- Cost of Room Service : " + payment.GetRsCharges());
+		System.out.println("- Tax : " + payment.GetTax() + "%");
+		System.out.println("- Total : " + payment.GetTotalBill());
+		
```
Then have to set the payment status to pending

sangkim98 committed:
update payment class

- almost complete javadoc
- more clear variable/method names
- detect input error in discount rate and tax rate

sangkim98 committed:
Completed Payment without input error management.

- Removed error handling in Payment class.
- Implemented `GenerateAndPrintBill()`

sangkim98 committed:
completed javadoc

- complete javadoc
- removed Scanner from `GenerateAndPrintBill(...)`

sangkim98 committed:
Specified data type in ArrayList

- Specified data type in ArrayList of `InitializeDB`

stephwongwt commented on src/main/payment/main/G1HRPS/PaymentManager.java:
```diff
@@ -152,43 +151,66 @@ public PaymentStatus GetPaymentStatus(Payment payment) {
+			price_of_order = each_order.CalTotalPrice();
+			System.out.println("--- Room Service Order [" + index + "]:");
+
+			// for(MenuItem menu : each_order.GetOrderedItemList()) {
```
remove code if not needed, if it's for future implementation, then leave a comment

sangkim98 committed:
Added CreateNewPayment(payment, class, attributes)

- Added CreateNewPayment method to create and add new payment to list

stephwongwt commented on src/main/payment/main/G1HRPS/PaymentManager.java:
```diff
@@ -184,7 +205,9 @@ public Payment GenerateAndPrintBill(String guest_id, int room_num, int days_of_s
-
+                        
+			// Prints all menu items in each order
+			
```
Hm please add a // TODO: as a reminder that this needs be changed. then the below code can be removed

yeohongwei reviewed (approved):
You might want to take a look at my RoomServiceOrder class on the function MenuItemtoString, so you may amend the printout for the GenerateAndPrintBill() later on if that's found to be useful.  I have no issue with the merging.

sangkim98 committed:
Print ordered items using menuItemstoString

- Used `menuItemstoString()` method of RoomServiceOrder class to print
ordered items in each order.
- Removed unnecessary comments in `InitializeDB()`

sangkim98 commented:
> You might want to take a look at my RoomServiceOrder class on the function MenuItemtoString, so you may amend the printout for the GenerateAndPrintBill() later on if that's found to be useful. I have no issue with the merging.

I've incorporated it in my code!

sangkim98 committed:
Merge branch 'main' into feature/payment
